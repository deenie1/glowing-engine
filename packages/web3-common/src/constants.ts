// Response error
export const ERR_RESPONSE = 100;
export const ERR_INVALID_RESPONSE = 101;

// Generic errors
export const ERR_PARAM = 200;
export const ERR_FORMATTERS = 201;
export const ERR_METHOD_NOT_IMPLEMENTED = 202;
export const ERR_OPERATION_TIMEOUT = 203;
export const ERR_OPERATION_ABORT = 204;

// Contract error codes
export const ERR_CONTRACT = 300;
export const ERR_CONTRACT_RESOLVER_MISSING = 301;
export const ERR_CONTRACT_ABI_MISSING = 302;
export const ERR_CONTRACT_REQUIRED_CALLBACK = 303;
export const ERR_CONTRACT_EVENT_NOT_EXISTS = 304;
export const ERR_CONTRACT_RESERVED_EVENT = 305;
export const ERR_CONTRACT_MISSING_DEPLOY_DATA = 306;
export const ERR_CONTRACT_MISSING_ADDRESS = 307;
export const ERR_CONTRACT_MISSING_FROM_ADDRESS = 308;

// Transaction error codes
export const ERR_TX = 400;
export const ERR_TX_REVERT_INSTRUCTION = 401;
export const ERR_TX_REVERT_TRANSACTION = 402;
export const ERR_TX_NO_CONTRACT_ADDRESS = 403;
export const ERR_TX_CONTRACT_NOT_STORED = 404;
export const ERR_TX_REVERT_WITHOUT_REASON = 405;
export const ERR_TX_OUT_OF_GAS = 406;

// Connection error codes
export const ERR_CONN = 500;
export const ERR_CONN_INVALID = 501;
export const ERR_CONN_TIMEOUT = 502;
export const ERR_CONN_NOT_OPEN = 503;
export const ERR_CONN_CLOSE = 504;
export const ERR_CONN_MAX_ATTEMPTS = 505;
export const ERR_CONN_PENDING_REQUESTS = 506;

// Provider error codes
export const ERR_PROVIDER = 600;
export const ERR_INVALID_PROVIDER = 601;
export const ERR_INVALID_CLIENT = 602;
export const ERR_SUBSCRIPTION = 603;

export const GENESIS_BLOCK_NUMBER = '0x0';
