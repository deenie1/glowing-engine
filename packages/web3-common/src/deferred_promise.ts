/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

import { OperationTimeoutError } from 'web3-errors';

export const promiseTimeout = async <T = void>(ms: number, message: string): Promise<T> =>
	new Promise((_, reject) => {
		const id = setTimeout(() => {
			clearTimeout(id);
			reject(new OperationTimeoutError(message ?? `Timed out in ${ms}ms.`));
		}, ms);
	});

export class DeferredPromise<T> implements Promise<T> {
	// public tag to treat object as promise by different libs
	public [Symbol.toStringTag]: 'Promise';

	private readonly _promise: Promise<T>;
	private _resolve!: (value: T | PromiseLike<T>) => void;
	private _reject!: (reason?: unknown) => void;
	private _state: 'pending' | 'fulfilled' | 'rejected' = 'pending';

	public constructor(
		{
			timeout,
			eagerStart,
			timeoutMessage,
		}: { timeout: number; eagerStart: boolean; timeoutMessage: string } = {
			timeout: 0,
			eagerStart: false,
			timeoutMessage: 'DeferredPromise timed out',
		},
	) {
		if (eagerStart) {
			this._promise = Promise.race([
				new Promise<T>((resolve, reject) => {
					this._resolve = resolve;
					this._reject = reject;
				}),
				promiseTimeout(timeout, timeoutMessage),
			]) as Promise<T>;
		} else {
			this._promise = new Promise<T>((resolve, reject) => {
				this._resolve = resolve;
				this._reject = reject;
			});
		}
	}

	public get state(): 'pending' | 'fulfilled' | 'rejected' {
		return this._state;
	}

	public async then<TResult1, TResult2>(
		onfulfilled?: (value: T) => TResult1 | PromiseLike<TResult1>,
		onrejected?: (reason: unknown) => TResult2 | PromiseLike<TResult2>,
	): Promise<TResult1 | TResult2> {
		return this._promise.then(onfulfilled, onrejected);
	}

	public async catch<TResult>(
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		onrejected?: (reason: any) => TResult | PromiseLike<TResult>,
	): Promise<T | TResult> {
		return this._promise.catch(onrejected);
	}

	public async finally(onfinally?: (() => void) | null): Promise<T> {
		return this._promise.finally(onfinally);
	}

	public resolve(value: T | PromiseLike<T>): void {
		this._state = 'fulfilled';
		this._resolve(value);
	}

	public reject(reason?: unknown): void {
		this._state = 'rejected';
		this._reject(reason);
	}
}
