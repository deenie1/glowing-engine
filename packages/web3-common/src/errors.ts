/* eslint-disable max-classes-per-file */

import {
	ERR_RESPONSE,
	ERR_PARAM,
	ERR_METHOD_NOT_IMPLEMENTED,
	ERR_CONN,
	ERR_CONN_INVALID,
	ERR_CONN_TIMEOUT,
	ERR_CONN_NOT_OPEN,
	ERR_CONN_CLOSE,
	ERR_CONN_MAX_ATTEMPTS,
	ERR_CONN_PENDING_REQUESTS,
	ERR_INVALID_PROVIDER,
	ERR_INVALID_RESPONSE,
	ERR_TX_REVERT_INSTRUCTION,
	ERR_TX_REVERT_TRANSACTION,
	ERR_TX,
	ERR_TX_NO_CONTRACT_ADDRESS,
	ERR_TX_CONTRACT_NOT_STORED,
	ERR_TX_REVERT_WITHOUT_REASON,
	ERR_TX_OUT_OF_GAS,
	ERR_CONTRACT_RESOLVER_MISSING,
	ERR_CONTRACT_ABI_MISSING,
	ERR_CONTRACT_REQUIRED_CALLBACK,
	ERR_CONTRACT_EVENT_NOT_EXISTS,
	ERR_CONTRACT_RESERVED_EVENT,
	ERR_CONTRACT_MISSING_DEPLOY_DATA,
	ERR_CONTRACT_MISSING_ADDRESS,
	ERR_CONTRACT_MISSING_FROM_ADDRESS,
	ERR_FORMATTERS,
	ERR_INVALID_CLIENT,
	ERR_PROVIDER,
	ERR_SUBSCRIPTION,
	ERR_OPERATION_TIMEOUT,
	ERR_OPERATION_ABORT,
	ERR_ABI_ENCODING,
	ERR_INVALID_PRIVATE_KEY,
	ERR_PRIVATE_KEY_LENGTH,
} from './constants';
import { isResponseWithError } from './json_rpc';

import { ConnectionEvent, JsonRpcResponse, Receipt } from './types';

export abstract class Web3Error extends Error {
	public readonly name: string;
	public abstract readonly code: number;

	public constructor(msg: string) {
		super(msg);
		this.name = this.constructor.name;
		Error.captureStackTrace(this, Web3Error);
	}

	public toJSON() {
		return { name: this.name, code: this.code, message: this.message };
	}
}

export class InvalidNumberOfParamsError extends Web3Error {
	public code = ERR_PARAM;

	public constructor(public got: number, public expected: number, public method: string) {
		super(`Invalid number of parameters for "${method}". Got "${got}" expected "${expected}"!`);
	}

	public toJSON() {
		return {
			...super.toJSON(),
			got: this.got,
			expected: this.expected,
			method: this.method,
		};
	}
}

export class ConnectionError extends Web3Error {
	public code = ERR_CONN;
	public errorCode?: number;
	public errorReason?: string;

	public constructor(message: string, event?: ConnectionEvent) {
		super(message);

		if (event) {
			this.errorCode = event.code;
			this.errorReason = event.reason;
		}
	}

	public toJSON() {
		return { ...super.toJSON(), errorCode: this.errorCode, errorReason: this.errorReason };
	}
}

export class InvalidConnectionError extends ConnectionError {
	public constructor(public host: string, event?: ConnectionEvent) {
		super(`CONNECTION ERROR: Couldn't connect to node ${host}.`, event);
		this.code = ERR_CONN_INVALID;
	}

	public toJSON() {
		return { ...super.toJSON(), host: this.host };
	}
}

export class ConnectionTimeoutError extends ConnectionError {
	public constructor(public duration: number) {
		super(`CONNECTION TIMEOUT: timeout of ${duration}ms achieved`);
		this.code = ERR_CONN_TIMEOUT;
	}

	public toJSON() {
		return { ...super.toJSON(), duration: this.duration };
	}
}

export class ConnectionNotOpenError extends ConnectionError {
	public constructor(event?: ConnectionEvent) {
		super('Connection not open', event);
		this.code = ERR_CONN_NOT_OPEN;
	}
}

export class ConnectionCloseError extends ConnectionError {
	public constructor(event?: ConnectionEvent) {
		super(
			`CONNECTION ERROR: The connection got closed with the close code ${
				event?.code ?? ''
			} and the following reason string ${event?.reason ?? ''}`,
			event,
		);
		this.code = ERR_CONN_CLOSE;
	}
}

export class MaxAttemptsReachedOnReconnectingError extends ConnectionError {
	public constructor() {
		super('Maximum number of reconnect attempts reached!');
		this.code = ERR_CONN_MAX_ATTEMPTS;
	}
}

export class PendingRequestsOnReconnectingError extends ConnectionError {
	public constructor() {
		super('CONNECTION ERROR: Provider started to reconnect before the response got received!');
		this.code = ERR_CONN_PENDING_REQUESTS;
	}
}

export class ProviderError extends Web3Error {
	public code = ERR_PROVIDER;
}

export class InvalidProviderError extends Web3Error {
	public code = ERR_INVALID_PROVIDER;

	public constructor(public clientUrl: string) {
		super(`Provider with url "${clientUrl}" is not set or invalid`);
	}
}

export class SubscriptionError extends Web3Error {
	public code = ERR_SUBSCRIPTION;
}

const buildErrorMessage = (response: JsonRpcResponse<unknown, unknown>): string =>
	isResponseWithError(response) ? response.error.message : '';

export class ResponseError<ErrorType = unknown> extends Web3Error {
	public code = ERR_RESPONSE;
	public data?: ErrorType | ErrorType[];

	public constructor(response: JsonRpcResponse<unknown, ErrorType>, message?: string) {
		super(
			message ??
				`Returned error: ${
					Array.isArray(response)
						? response.map(r => buildErrorMessage(r)).join(',')
						: buildErrorMessage(response)
				}`,
		);

		if (!message) {
			this.data = Array.isArray(response)
				? response.map(r => r.error?.data as ErrorType)
				: response?.error?.data;
		}
	}

	public toJSON() {
		return { ...super.toJSON(), data: this.data };
	}
}

export class InvalidResponseError<ErrorType = unknown> extends ResponseError<ErrorType> {
	public constructor(result: JsonRpcResponse<unknown, ErrorType>) {
		super(result);
		if (!this.message || this.message === '') {
			this.message = `Invalid JSON RPC response: ${JSON.stringify(result)}`;
		}
		this.code = ERR_INVALID_RESPONSE;
	}
}

export class TransactionError extends Web3Error {
	public code = ERR_TX;

	public constructor(message: string, public receipt: Receipt) {
		super(message);
	}

	public toJSON() {
		return { ...super.toJSON(), receipt: this.receipt };
	}
}

export class RevertInstructionError extends Web3Error {
	public code = ERR_TX_REVERT_INSTRUCTION;

	public constructor(public reason: string, public signature: string) {
		super(`Your request got reverted with the following reason string: ${reason}`);
	}

	public toJSON() {
		return { ...super.toJSON(), reason: this.reason, signature: this.signature };
	}
}

export class TransactionRevertError extends Web3Error {
	public code = ERR_TX_REVERT_TRANSACTION;

	public constructor(public reason: string, public signature: string, public receipt: Receipt) {
		super(`Transaction has been reverted by the EVM:\n ${JSON.stringify(receipt, null, 2)}`);
	}

	public toJSON() {
		return {
			...super.toJSON(),
			reason: this.reason,
			signature: this.signature,
			receipt: this.receipt,
		};
	}
}

export class NoContractAddressFoundError extends TransactionError {
	public constructor(receipt: Receipt) {
		super("The transaction receipt didn't contain a contract address.", receipt);
		this.code = ERR_TX_NO_CONTRACT_ADDRESS;
	}

	public toJSON() {
		return { ...super.toJSON(), receipt: this.receipt };
	}
}

export class ContractCodeNotStoredError extends TransactionError {
	public constructor(receipt: Receipt) {
		super("The contract code couldn't be stored, please check your gas limit.", receipt);
		this.code = ERR_TX_CONTRACT_NOT_STORED;
	}
}

export class TransactionRevertedWithoutReasonError extends TransactionError {
	public constructor(receipt: Receipt) {
		super(
			`Transaction has been reverted by the EVM:\n ${JSON.stringify(receipt, null, 2)}`,
			receipt,
		);
		this.code = ERR_TX_REVERT_WITHOUT_REASON;
	}
}

export class TransactionOutOfGasError extends TransactionError {
	public constructor(receipt: Receipt) {
		super(
			`Transaction ran out of gas. Please provide more gas:\n ${JSON.stringify(
				receipt,
				null,
				2,
			)}`,
			receipt,
		);
		this.code = ERR_TX_OUT_OF_GAS;
	}
}

export class ResolverMethodMissingError extends Web3Error {
	public code = ERR_CONTRACT_RESOLVER_MISSING;

	public constructor(public address: string, public name: string) {
		super(`The resolver at ${address} does not implement requested method: "${name}".`);
	}

	public toJSON() {
		return { ...super.toJSON(), address: this.address, name: this.name };
	}
}

export class ContractMissingABIError extends Web3Error {
	public code = ERR_CONTRACT_ABI_MISSING;

	public constructor() {
		super(
			'You must provide the json interface of the contract when instantiating a contract object.',
		);
	}
}

export class ContractOnceRequiresCallbackError extends Web3Error {
	public code = ERR_CONTRACT_REQUIRED_CALLBACK;

	public constructor() {
		super('Once requires a callback as the second parameter.');
	}
}

export class ContractEventDoesNotExistError extends Web3Error {
	public code = ERR_CONTRACT_EVENT_NOT_EXISTS;

	public constructor(public eventName: string) {
		super(`Event "${eventName}" doesn't exist in this contract.`);
	}

	public toJSON() {
		return { ...super.toJSON(), eventName: this.eventName };
	}
}

export class ContractReservedEventError extends Web3Error {
	public code = ERR_CONTRACT_RESERVED_EVENT;

	public constructor(public type: string) {
		super(`Event "${type}" doesn't exist in this contract.`);
	}

	public toJSON() {
		return { ...super.toJSON(), type: this.type };
	}
}

export class ContractMissingDeployDataError extends Web3Error {
	public code = ERR_CONTRACT_MISSING_DEPLOY_DATA;

	public constructor() {
		super(`No "data" specified in neither the given options, nor the default options.`);
	}
}

export class ContractNoAddressDefinedError extends Web3Error {
	public code = ERR_CONTRACT_MISSING_ADDRESS;

	public constructor() {
		super("This contract object doesn't have address set yet, please set an address first.");
	}
}

export class ContractNoFromAddressDefinedError extends Web3Error {
	public code = ERR_CONTRACT_MISSING_FROM_ADDRESS;

	public constructor() {
		super('No "from" address specified in neither the given options, nor the default options.');
	}
}

export class FormatterError extends Web3Error {
	public code = ERR_FORMATTERS;
}

export class MethodNotImplementedError extends Web3Error {
	public code = ERR_METHOD_NOT_IMPLEMENTED;

	public constructor() {
		super("The method you're trying to call is not implemented.");
	}
}

export class InvalidClientError extends Web3Error {
	public code = ERR_INVALID_CLIENT;

	public constructor(clientUrl: string) {
		super(`Client URL "${clientUrl}" is invalid.`);
	}
}

export class OperationTimeoutError extends Web3Error {
	public code = ERR_OPERATION_TIMEOUT;
}

export class OperationAbortError extends Web3Error {
	public code = ERR_OPERATION_ABORT;
}

export class AbiError extends Web3Error {
	public code = ERR_ABI_ENCODING;
}

export class PrivateKeyLengthError extends Web3Error {
	public code = ERR_PRIVATE_KEY_LENGTH;
	public constructor(value: string | Buffer) {
		super(`Invalid value given "${value.toString()}". Error: Private key must be 32 bytes.`);
	}
}

export class InvalidPrivateKeyError extends Web3Error {
	public code = ERR_INVALID_PRIVATE_KEY;
	public constructor(value: string | Buffer) {
		super(`Invalid value given "${String(value)}". Error: not a valid string or buffer.`);
	}
}
