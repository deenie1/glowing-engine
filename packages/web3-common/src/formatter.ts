/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
import { FormatterError } from 'web3-errors';
import {
	Bytes,
	bytesToBuffer,
	bytesToHex,
	HexString,
	mergeDeep,
	Numbers,
	numberToHex,
	toBigInt,
} from 'web3-utils';
import { isNullish, isObject, JsonSchema, utils, ValidationSchemaInput } from 'web3-validator';

const { parseBaseType } = utils;

export enum FMT_NUMBER {
	NUMBER = 'NUMBER_NUMBER',
	HEX = 'NUMBER_HEX',
	STR = 'NUMBER_STR',
	BIGINT = 'NUMBER_BIGINT',
}

export type NumberTypes = {
	[FMT_NUMBER.NUMBER]: number;
	[FMT_NUMBER.HEX]: HexString;
	[FMT_NUMBER.STR]: string;
	[FMT_NUMBER.BIGINT]: bigint;
};

export enum FMT_BYTES {
	HEX = 'BYTES_HEX',
	BUFFER = 'BYTES_BUFFER',
	UINT8ARRAY = 'BYTES_UINT8ARRAY',
}

export type ByteTypes = {
	[FMT_BYTES.HEX]: HexString;
	[FMT_BYTES.BUFFER]: Buffer;
	[FMT_BYTES.UINT8ARRAY]: Uint8Array;
};

export type DataFormat = {
	readonly number: FMT_NUMBER;
	readonly bytes: FMT_BYTES;
};

export const DEFAULT_RETURN_FORMAT = { number: FMT_NUMBER.BIGINT, bytes: FMT_BYTES.HEX } as const;
export const ETH_DATA_FORMAT = { number: FMT_NUMBER.HEX, bytes: FMT_BYTES.HEX } as const;

export type FormatType<T, F extends DataFormat> = number extends Extract<T, Numbers>
	? NumberTypes[F['number']] | Exclude<T, Numbers>
	: Buffer extends Extract<T, Bytes>
	? ByteTypes[F['bytes']] | Exclude<T, Bytes>
	: T extends object | undefined
	? {
			[P in keyof T]: FormatType<T[P], F>;
	  }
	: T;

const findSchemaByDataPath = (schema: JsonSchema, dataPath: string[]): JsonSchema | undefined => {
	let result: JsonSchema = { ...schema } as JsonSchema;

	for (const dataPart of dataPath) {
		if (!result.properties && !result.items) {
			return undefined;
		}

		if (result.properties) {
			result = (result.properties as Record<string, JsonSchema>)[dataPart];
		} else if (result.items && (result.items as JsonSchema).properties) {
			const node = (result.items as JsonSchema).properties as Record<string, JsonSchema>;

			if (!node) {
				return undefined;
			}

			result = node[dataPart];
		} else if (result.items && isObject(result.items)) {
			result = result.items;
		} else if (result.items && Array.isArray(result.items)) {
			result = (result.items as JsonSchema[])[parseInt(dataPart, 10)];
		}
	}

	return result;
};

export const convertScalarValue = (value: unknown, ethType: string, format: DataFormat) => {
	try {
		const { baseType } = parseBaseType(ethType);

		if (baseType === 'int' || baseType === 'uint') {
			switch (format.number) {
				case FMT_NUMBER.NUMBER:
					return Number(toBigInt(value));
				case FMT_NUMBER.HEX:
					return numberToHex(toBigInt(value));
				case FMT_NUMBER.STR:
					return toBigInt(value).toString();
				case FMT_NUMBER.BIGINT:
					return toBigInt(value);
				default:
					throw new FormatterError(`Invalid format: ${String(format.number)}`);
			}
		}

		if (baseType === 'bytes') {
			switch (format.bytes) {
				case FMT_BYTES.HEX:
					return bytesToHex(bytesToBuffer(value as Bytes));
				case FMT_BYTES.BUFFER:
					return bytesToBuffer(value as Bytes);
				case FMT_BYTES.UINT8ARRAY:
					return new Uint8Array(bytesToBuffer(value as Bytes));
				default:
					throw new FormatterError(`Invalid format: ${String(format.bytes)}`);
			}
		}
	} catch (error) {
		// If someone did't used `eth` keyword we can return original value
		// as the scope of this code is formatting not validation
		return value;
	}

	return value;
};

export const convert = (
	data: Record<string, unknown> | unknown[] | unknown,
	schema: JsonSchema,
	dataPath: string[],
	format: DataFormat,
) => {
	// If it's a scalar value
	if (!isObject(data) && !Array.isArray(data)) {
		return convertScalarValue(data, schema?.eth as string, format);
	}

	const object = data as Record<string, unknown>;

	for (const [key, value] of Object.entries(object)) {
		dataPath.push(key);
		const schemaProp = findSchemaByDataPath(schema, dataPath);

		// If value is a scaler value
		if (isNullish(schemaProp)) {
			delete object[key];
			dataPath.pop();

			continue;
		}

		// If value is an object, recurse into it
		if (isObject(value)) {
			convert(value, schema, dataPath, format);
			dataPath.pop();
			continue;
		}

		// If value is an array
		if (Array.isArray(value)) {
			if (isNullish(schemaProp?.items)) {
				// Can not find schema for array item, delete that item
				delete object[key];
				dataPath.pop();

				continue;
			}

			// If schema for array items is a single type
			if (isObject(schemaProp.items) && !isNullish(schemaProp.items.eth)) {
				for (let i = 0; i < value.length; i += 1) {
					(object[key] as unknown[])[i] = convertScalarValue(
						value[i],
						// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
						schemaProp?.items?.eth as string,
						format,
					);
				}

				dataPath.pop();
				continue;
			}

			// If schema for array items is an object
			if (
				!Array.isArray(schemaProp?.items) &&
				(schemaProp?.items as JsonSchema).type === 'object'
			) {
				for (const arrObject of value) {
					convert(
						arrObject as Record<string, unknown> | unknown[],
						schema,
						dataPath,
						format,
					);
				}

				dataPath.pop();
				continue;
			}

			// If schema for array is a tuple
			if (Array.isArray(schemaProp?.items)) {
				for (let i = 0; i < value.length; i += 1) {
					(object[key] as unknown[])[i] = convertScalarValue(
						value[i],
						(schemaProp.items as JsonSchema[])[i].eth as string,
						format,
					);
				}

				dataPath.pop();
				continue;
			}
		}

		object[key] = convertScalarValue(value, schemaProp.eth as string, format);

		dataPath.pop();
	}

	return object;
};

export const format = <
	DataType extends Record<string, unknown> | unknown[] | unknown,
	ReturnType extends DataFormat,
>(
	schema: ValidationSchemaInput | JsonSchema,
	data: DataType,
	returnFormat: ReturnType,
): FormatType<DataType, ReturnType> => {
	let dataToParse: Record<string, unknown> | unknown[] | unknown;

	if (isObject(data)) {
		dataToParse = mergeDeep({}, data);
	} else if (Array.isArray(data)) {
		dataToParse = [...data];
	} else {
		dataToParse = data;
	}

	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
	const jsonSchema: JsonSchema = isObject(schema) ? schema : utils.ethAbiToJsonSchema(schema);

	if (!jsonSchema.properties && !jsonSchema.items && !jsonSchema.eth) {
		throw new FormatterError('Invalid json schema for formatting');
	}

	return convert(dataToParse, jsonSchema, [], returnFormat) as FormatType<
		typeof data,
		ReturnType
	>;
};
