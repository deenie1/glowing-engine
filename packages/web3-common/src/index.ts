export * from './errors';
export * from './constants';
export * from './types';
export * from './web3_base_provider';
export * from './web3_base_wallet';
export * from './web3_event_emitter';
export * from './eth_execution_api';
export * from './deferred_promise';
export * as jsonRpc from './json_rpc';
