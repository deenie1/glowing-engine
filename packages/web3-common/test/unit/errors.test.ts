/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

import * as errors from '../../src/index';

describe('errors', () => {
	describe('ResponseError', () => {
		it('should have valid json structure with data', () => {
			expect(
				new errors.ResponseError({
					id: 1,
					jsonrpc: '2.0',
					error: { code: 123, message: 'error message', data: { a: '10', b: '20' } },
				}).toJSON(),
			).toMatchSnapshot();
		});

		it('should have valid json structure without data', () => {
			expect(
				new errors.ResponseError({
					id: 1,
					jsonrpc: '2.0',
					error: { code: 123, message: 'error message', data: undefined },
				}).toJSON(),
			).toMatchSnapshot();
		});
	});

	describe('InvalidResponseError', () => {
		it('should have valid json structure', () => {
			expect(
				new errors.InvalidResponseError({
					id: 1,
					jsonrpc: '2.0',
					error: { code: 123, message: 'error message', data: { a: '10', b: '20' } },
				}).toJSON(),
			).toMatchSnapshot();
		});
	});
});
