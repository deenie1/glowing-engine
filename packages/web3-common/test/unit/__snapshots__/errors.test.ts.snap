// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors ConnectionCloseError should have valid json structure 1`] = `
Object {
  "code": 504,
  "errorCode": 10,
  "errorReason": "reason",
  "message": "CONNECTION ERROR: The connection got closed with the close code 10 and the following reason string reason",
  "name": "ConnectionCloseError",
}
`;

exports[`errors ConnectionError should have valid json structure 1`] = `
Object {
  "code": 500,
  "errorCode": 10,
  "errorReason": "reason",
  "message": "error message",
  "name": "ConnectionError",
}
`;

exports[`errors ConnectionNotOpenError should have valid json structure 1`] = `
Object {
  "code": 503,
  "errorCode": 10,
  "errorReason": "reason",
  "message": "Connection not open",
  "name": "ConnectionNotOpenError",
}
`;

exports[`errors ConnectionTimeoutError should have valid json structure 1`] = `
Object {
  "code": 502,
  "duration": 5000,
  "errorCode": undefined,
  "errorReason": undefined,
  "message": "CONNECTION TIMEOUT: timeout of 5000ms achieved",
  "name": "ConnectionTimeoutError",
}
`;

exports[`errors ContractCodeNotStoredError should have valid json structure 1`] = `
Object {
  "code": 404,
  "message": "The contract code couldn't be stored, please check your gas limit.",
  "name": "ContractCodeNotStoredError",
  "receipt": Object {
    "attr1": "attr1",
  },
}
`;

exports[`errors ContractEventDoesNotExistError should have valid json structure 1`] = `
Object {
  "code": 304,
  "eventName": "eventName",
  "message": "Event \\"eventName\\" doesn't exist in this contract.",
  "name": "ContractEventDoesNotExistError",
}
`;

exports[`errors ContractMissingABIError should have valid json structure 1`] = `
Object {
  "code": 302,
  "message": "You must provide the json interface of the contract when instantiating a contract object.",
  "name": "ContractMissingABIError",
}
`;

exports[`errors ContractMissingDeployDataError should have valid json structure 1`] = `
Object {
  "code": 306,
  "message": "No \\"data\\" specified in neither the given options, nor the default options.",
  "name": "ContractMissingDeployDataError",
}
`;

exports[`errors ContractNoAddressDefinedError should have valid json structure 1`] = `
Object {
  "code": 307,
  "message": "This contract object doesn't have address set yet, please set an address first.",
  "name": "ContractNoAddressDefinedError",
}
`;

exports[`errors ContractNoFromAddressDefinedError should have valid json structure 1`] = `
Object {
  "code": 308,
  "message": "No \\"from\\" address specified in neither the given options, nor the default options.",
  "name": "ContractNoFromAddressDefinedError",
}
`;

exports[`errors ContractOnceRequiresCallbackError should have valid json structure 1`] = `
Object {
  "code": 303,
  "message": "Once requires a callback as the second parameter.",
  "name": "ContractOnceRequiresCallbackError",
}
`;

exports[`errors ContractReservedEventError should have valid json structure 1`] = `
Object {
  "code": 305,
  "message": "Event \\"type\\" doesn't exist in this contract.",
  "name": "ContractReservedEventError",
  "type": "type",
}
`;

exports[`errors InvalidConnectionError should have valid json structure 1`] = `
Object {
  "code": 501,
  "errorCode": 10,
  "errorReason": "reason",
  "host": "my host",
  "message": "CONNECTION ERROR: Couldn't connect to node my host.",
  "name": "InvalidConnectionError",
}
`;

exports[`errors InvalidNumberOfParamsError should have valid json structure 1`] = `
Object {
  "code": 200,
  "expected": 20,
  "got": 10,
  "message": "Invalid number of parameters for \\"method\\". Got \\"10\\" expected \\"20\\"!",
  "method": "method",
  "name": "InvalidNumberOfParamsError",
}
`;

exports[`errors InvalidProviderError should have valid json structure 1`] = `
Object {
  "code": 601,
  "message": "Provider with url \\"my url\\" is not set or invalid",
  "name": "InvalidProviderError",
}
`;

exports[`errors InvalidResponseError should have valid json structure 1`] = `
Object {
  "code": 101,
  "data": Object {
    "a": "10",
    "b": "20",
  },
  "message": "Returned error: error message",
  "name": "InvalidResponseError",
}
`;

exports[`errors MaxAttemptsReachedOnReconnectingError should have valid json structure 1`] = `
Object {
  "code": 505,
  "errorCode": undefined,
  "errorReason": undefined,
  "message": "Maximum number of reconnect attempts reached!",
  "name": "MaxAttemptsReachedOnReconnectingError",
}
`;

exports[`errors NoContractAddressFoundError should have valid json structure 1`] = `
Object {
  "code": 403,
  "message": "The transaction receipt didn't contain a contract address.",
  "name": "NoContractAddressFoundError",
  "receipt": Object {
    "attr1": "attr1",
  },
}
`;

exports[`errors PendingRequestsOnReconnectingError should have valid json structure 1`] = `
Object {
  "code": 506,
  "errorCode": undefined,
  "errorReason": undefined,
  "message": "CONNECTION ERROR: Provider started to reconnect before the response got received!",
  "name": "PendingRequestsOnReconnectingError",
}
`;

exports[`errors ResolverMethodMissingError should have valid json structure 1`] = `
Object {
  "address": "address",
  "code": 301,
  "message": "The resolver at address does not implement requested method: \\"name\\".",
  "name": "name",
}
`;

exports[`errors ResponseError should have valid json structure with data 1`] = `
Object {
  "code": 100,
  "data": Object {
    "a": "10",
    "b": "20",
  },
  "message": "Returned error: error message",
  "name": "ResponseError",
}
`;

exports[`errors ResponseError should have valid json structure without data 1`] = `
Object {
  "code": 100,
  "data": undefined,
  "message": "Returned error: error message",
  "name": "ResponseError",
}
`;

exports[`errors RevertInstructionError should have valid json structure 1`] = `
Object {
  "code": 401,
  "message": "Your request got reverted with the following reason string: message",
  "name": "RevertInstructionError",
  "reason": "message",
  "signature": "signature",
}
`;

exports[`errors TransactionError should have valid json structure 1`] = `
Object {
  "code": 400,
  "message": "message",
  "name": "TransactionError",
  "receipt": Object {
    "attr1": "attr1",
  },
}
`;

exports[`errors TransactionOutOfGasError should have valid json structure 1`] = `
Object {
  "code": 406,
  "message": "Transaction ran out of gas. Please provide more gas:
 {
  \\"attr1\\": \\"attr1\\"
}",
  "name": "TransactionOutOfGasError",
  "receipt": Object {
    "attr1": "attr1",
  },
}
`;

exports[`errors TransactionRevertError should have valid json structure 1`] = `
Object {
  "code": 402,
  "message": "Transaction has been reverted by the EVM:
 {
  \\"attr1\\": \\"attr1\\"
}",
  "name": "TransactionRevertError",
  "reason": "message",
  "receipt": Object {
    "attr1": "attr1",
  },
  "signature": "signature",
}
`;

exports[`errors TransactionRevertedWithoutReasonError should have valid json structure 1`] = `
Object {
  "code": 405,
  "message": "Transaction has been reverted by the EVM:
 {
  \\"attr1\\": \\"attr1\\"
}",
  "name": "TransactionRevertedWithoutReasonError",
  "receipt": Object {
    "attr1": "attr1",
  },
}
`;
